### PR TITLE
BatchMaxQueries

### PR DESCRIPTION
**Pull request by happylinks**
Issue reference: #1654 

TODO:

- [X] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [X] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [ ] If this was a change that affects the external API, update the docs and post a link to the PR in the discussion

### CODE REVIEW

**Comment by happylinks** [comment]
@helfer This is my first try. I'm having some issues with creating a correct test for this. Any hints would be great. I'll also keep trying myself of course.
After that I will add documentation, if you think this is the way to go.

File: src/transport/batching.ts [hunk 1]
```diff
@@ -62,20 +66,20 @@ export class QueryBatcher {
   // Consumes the queue.
   // Returns a list of promises (one for each query).
   public consumeQueue(): (Promise<ExecutionResult> | undefined)[] | undefined {
-    const requests: Request[] = this.queuedRequests.map(
+    const queueSlice = this.queuedRequests.splice(0, this.batchMax);
```
Comment (helfer) [inline src/transport/batching.ts]: Instead of slicing up when the queue is consumed, it would be better to immediately consume the queue once it reaches a length of `batchMax`. Otherwise you're still going to send all of those queries at the same time, just in chunks.

**Comment by happylinks** [comment]
Sorry for the slow response. I actually got it working now, thanks for the hints. Do you think this test covers the full use-case or should I make more?
Also, should this be a breaking change, because if you don't have batchMax in your settings it will throw an error. This was also the case for batchInterval, but I'm not sure what the reasoning for that was. We could have a default maybe?

**Comment by happylinks** [comment]
Hi @helfer, I see you tried to merge master again but now the tests fail. Is there something I can do to help speed this PR up :)?

File: src/transport/batching.ts [hunk 2]
```diff
@@ -24,20 +24,24 @@ export class QueryBatcher {
...
-  private batchInterval: Number;
+  private batchInterval: number;
+  private batchMax: number;
 
   //This function is called to the queries in the queue to the server.
   private batchFetchFunction: (request: Request[]) => Promise<ExecutionResult[]>;
 
   constructor({
     batchInterval,
+    batchMax,
```
Comment (helfer) [inline src/transport/batching.ts]: Let's make `batchMax` optional, and set it to 0 by default, in which case it should behave the same way it did before this PR.

File: test/batching.ts [hunk 3]
```diff
@@ -16,6 +16,7 @@ describe('QueryBatcher', () => {
     assert.doesNotThrow(() => {
       const querySched = new QueryBatcher({
         batchInterval: 10,
+        batchMax: 10,
```
Comment (helfer) [inline test/batching.ts]: If we make it optional, we won't have to add it to every test. This will also ensure that the change is not breaking for current users.

File: test/client.ts [hunk 4]
```diff
@@ -2201,6 +2204,109 @@ describe('client', () => {
     });
   });
 
+  it('should limit the amount of queries in a batch according to the batchMax value', (done) => {
```
Comment (helfer) [inline test/client.ts]: This test looks good! After making `batchMax` optional, please also add a test that checks that no batching happens if `batchMax` is set to zero.

**Comment by happylinks** [comment]
Hi @helfer, I made it optional. Also I changed the params of the HTTPBatchNetworkInterface to be an object. This made it nicer to have optional params. Let me know if that's ok.
I also made a PR for the documentation. Not sure if I should add it to more pages, but this seemed the best place.

**Comment by helfer** [comment]
It looks great now, thanks a lot @happylinks!

**Comment by happylinks** [comment]
Awesome! Thanks for the help :)

**Comment by jaydenseric** [comment]
Just a heads up that the [change in the constructor](https://github.com/apollographql/apollo-client/pull/1659/files#diff-e835769b5317203454f623b7ffad4772L49) from parameters to an object is [a breaking change](https://github.com/jaydenseric/apollo-upload-client/blob/v4.1.0/src/batch-network-interface.js#L52) for libraries like `apollo-upload-client`. After people update from v1.4.2 to v1.5.0, the error `A remote endpoint is required for a network layer` occurs. I'll quickly push out an update now.

I like the change to an object, it just would have been nice as a semver breaking change (`apollo-client@2.0.0`).

**Comment by helfer** [comment]
Hey @jaydenseric sorry about that!
